File: leetcode/js/6/zigZag.js
var convert = function(s, numRows) {
    if(s.length < numRows || s.length < 3 || numRows === 1) {
        return s;
    }

    let result = [];
    let string = s.split('');
    let col = 0;
    for(let i = 0; i < numRows; i++) {
        result.push([]);
    }

    result[0][0] = string.shift();

    while(string.length > 0) {
        for(let j = 1; j < numRows; j++) {
            result[j][col] = string.shift();
        }
    
        for(let k = numRows - 2; k >= 0; k--) {
            col++;
            result[k][col] = string.shift();
        }
    }

    return result.flat().join('');
};

File: leetcode/python/6/zigzag.py
class Solution(object):
    def fill(self, numRows, arr):
        col = len(arr)
        arr.append([])
        for x in range(numRows):
            arr[col].append('')

    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        # set up once
        arr = []
        z = 0
        i = 0

        s_length = len(s)

        while z < s_length:
            # pattern one
            self.fill(numRows, arr)
            y = 0
            while y < numRows:
                if z == s_length:
                    break
                arr[i][y] = s[z]
                y += 1
                z += 1

            # pattern two
            i += 1
            j = numRows - 2
            while j > 0:
                if z == s_length:
                    break
                self.fill(numRows, arr)
                arr[i][j] = s[z]
                i += 1
                j -= 1
                z += 1

        word = ''
        return word


sol = Solution()
sol.convert("PAYPALISHIRING", 3)